src/riscv/lib/src/jit/builder/control_flow_graph.rs

```diff
@@ -179,122 +184,91 @@ where
-        }
-    }
-
-    /// Find the step deltas for each node in the graph.
+    /// Initialise the outcome queue for finding step counter updates.
     ///
-    /// A step delta is the number of steps that have been performed since the last step counter
-    /// update. 0 means the step counter was updated through an incoming edge of a node.
-    fn find_step_deltas(&self) -> InstrMap<usize> {
-        let mut step_deltas = self.nodes.map(|_, _| None);
-
-        for (idx, _) in self.nodes.iter() {
-            let mut walker = GraphWalker::new(idx);
-            while let Some(cursor) = walker.next() {
-                let pos = cursor.position();
-
-                if step_deltas[pos].is_some() {
-                    // If the step delta is already set there is nothing for us to do.
-                    continue;
-                }
-
-                let EdgeClass::Unambiguous(source) = self.classify_incoming(pos) else {
-                    // Join points require their incoming edges to flush the step counter, thereby
-                    // setting delta back to 0. Unreachable nodes get set to 0 as well, as their
-                    // step delta is irrelevant.
-                    step_deltas[pos] = Some(0usize);
-                    continue;
-                };
-
-                if let Some(delta) = step_deltas[source] {
-                    // If the previous node already has a step delta, we can just use that to
-                    // determine the current node's step delta.
-                    step_deltas[pos] = Some(delta + 1);
-                    continue;
-                }
+    /// This will result in all entry points having an instruction delta of 0, and
+    /// all outcomes from those entry points being added to the queue.
+    fn init_outcome_queue(&self, step_deltas: &mut InstrMap<Option<usize>>) -> VecDeque<OutcomeId> {
+        let mut queue = VecDeque::new();
+
+        for idx in self.graph.outgoing_outcomes(SourceInstrLoc::Entry) {
+            let TargetInstrLoc::Internal(start_idx) = self.outcomes[*idx].to() else {
+                continue;
+            };
 
-                if cursor.already_seen(source) {
-                    // We have encountered a cycle. That means it is impossible to resolve an
-                    // accurate delta. We simply require the incoming edge to flush the step
-                    // counter. This makes the step delta 0.
-                    step_deltas[pos] = Some(0usize);
-                    continue;
-                }
+            step_deltas[start_idx] = Some(0);
 
-                cursor.not_done_yet([source]);
+            for outcome in self
+                .graph
+                .outgoing_outcomes(SourceInstrLoc::Internal(start_idx))
+            {
+                queue.push_back(*outcome);
             }
         }
 
-        step_deltas.map(|_idx, delta| delta.expect("All nodes should have a step delta by now"))
+        queue
     }
```

Comment:
The existence of this function seems odd. The algorithm we discussed doesn't call for an initial traversal. The rules applied for each transition should cover this case without the need for duplicate traversal.